<a href="https://colab.research.google.com/github/TD1138/Interactive-Movie-Recommender/blob/main/notebooks/Interactive_Movie_Recommender_TF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# An Interactive Movie Recommender using Tensorflow

### Most movie recommender systems I've seen before require a starting point.
### This can be in the form of:
* ###   a reference movie (if you enjoyed X you will enjoy Y or Z)
* ###   a taste profile ('you liked A, B & C so how about D or E')

### I have 2 problems with these kinds of recommender systems:
### If I ask you where you want to go for dinner, it's an open-ended question - one is often  unable to come up with something. However if I ask where you'd rather go from 3 options, that is a much easier question to answer! Once we have a starting point we can start to narrow things down. This is how I'd like a recommender system to work - give me some options and I'll be able to narrow things down!
### The second issue is the need for a starting point - I either need to think of a film that I want any recommendations to be similar to, or I need to have already filled out various films I enjoy. In reality, I sometimes don't know exactly what film I want a film to be similar to, I jsut know what kind of thing I'm after!


### So my idea is to make a recommender that is more interactive than a classic  - taking input from the user.


### For example:
#### - User inputs 'Sci-Fi'
#### * Algorithm outputs 10 'Sci-Fi movies'
#### - User sees that the movies skew quite modern - enters '80s'
#### * Algorithm outputs 10 '80s Sci-Fi movies'
#### - User sees most of these movies feature male leads - adds 'female lead'
#### * Algorithm outputs 10 '80s Sci-Fi movies with Female Leads'

### My idea is to create a neural network with inputs being various features like year, director etc, but also some more meta features, like the tags you see in Netflix's micro-genres - like 'gory' 'visualy striking' etc.
### The output of the neural network would be a softmax with all the films in our dataset. The top 10 probabilities would then be served up as the recommendations.
### It will require a layer of natural language processing to get from the user inputs to the feature space - this will be a second phase of this project!

### But before I can get started with this grand ambition, I need some data! 


# The Data

### GroupLens is a computer science research lab at the University of Minnesota, specializing in recommender systems, among other things.
### They have a dataset which is ideal for my usecase - 'MovieLens 25M'
### Released December 2019, it contains 62k movies, with tags like the ones I was talking about above! See further documentation below:
### https://grouplens.org/datasets/movielens/


### The data is hosted on the GroupLens website, as a zip file.
### First I need to save the ZIP file to the local machine, then use a package to extract the data:

In [1]:
dataset_url = 'http://files.grouplens.org/datasets/movielens/ml-25m.zip'

In [2]:
import os
local_path_zip = os.path.join('..', 'dataset/ml-25m.zip')
local_path = os.path.join('..', 'dataset/ml-25m.zip/ml-25m')

In [3]:
import requests, zipfile, io
r = requests.get(dataset_url)
z = zipfile.ZipFile(io.BytesIO(r.content))
z.extractall(local_path_zip)

####Let's have a look at the files we have downloaded:

In [147]:
pd.set_option("display.max_rows", 500, "display.max_columns", 50)
import pandas as pd
for file in os.listdir(local_path):
  if file[-4:] == '.csv':
    print(file)
    display(pd.read_csv(local_path+'/'+file, nrows=5))
  else:
    print(file+' is not a csv')
  print('\n')

genome-tags.csv


tagId           tag
0      1           007
1      2  007 (series)
2      3  18th century
3      4         1920s
4      5         1930s



links.csv


movieId  imdbId  tmdbId
0        1  114709     862
1        2  113497    8844
2        3  113228   15602
3        4  114885   31357
4        5  113041   11862



movies.csv


movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy



tags.csv


userId  movieId               tag   timestamp
0       3      260           classic  1439472355
1       3      260            sci-fi  1439472256
2       4     1732       dark comedy  1573943598
3       4     1732    great dialogue  1573943604
4       4     7569  so bad it's good  1573943455



genome-scores.csv


movieId  tagId  relevance
0        1      1    0.02875
1        1      2    0.02375
2        1      3    0.06250
3        1      4    0.07575
4        1      5    0.14075



README.txt is not a csv


ratings.csv


userId  movieId  rating   timestamp
0       1      296     5.0  1147880044
1       1      306     3.5  1147868817
2       1      307     5.0  1147868828
3       1      665     5.0  1147878820
4       1      899     3.5  1147868510

### Let's read in the 'movies.csv' file for later use, and do some basic processing of the data:

In [111]:
movies_df = pd.read_csv(local_path+'/movies.csv')
#movies_df = movies_df.head(10)
movies_df['year'] = movies_df['title'].str.split(' \(').str[1].str.replace(')','')
movies_df['title'] = movies_df['title'].str.split(' \(').str[0]
#movies_df['genre_count']
movies_df.head(10)

movieId                        title  \
0        1                    Toy Story   
1        2                      Jumanji   
2        3             Grumpier Old Men   
3        4            Waiting to Exhale   
4        5  Father of the Bride Part II   
5        6                         Heat   
6        7                      Sabrina   
7        8                 Tom and Huck   
8        9                 Sudden Death   
9       10                    GoldenEye   

                                        genres  year  
0  Adventure|Animation|Children|Comedy|Fantasy  1995  
1                   Adventure|Children|Fantasy  1995  
2                               Comedy|Romance  1995  
3                         Comedy|Drama|Romance  1995  
4                                       Comedy  1995  
5                        Action|Crime|Thriller  1995  
6                               Comedy|Romance  1995  
7                           Adventure|Children  1995  
8                                       Action  1995  
9                    Action|Adventure|Thriller  1995

### Let's do some splitting/tokenising of the genres:

In [137]:
genres_processing = movies_df['genres'].str.split('|', expand=True)
#genres_processing.columns = ['genre_'+str(i) for i in genres_processing.columns]
genres_processing
genres_list = []
for col in genres_processing.columns:
  genres_counts = genres_processing[col].value_counts()
  genres_list.append(genres_counts)

genres_table = pd.concat(genres_list).reset_index().groupby('index').sum()
genres_table.reset_index(inplace=True)
genres_table.columns = ['genre', 'movie_count']
genres_table = genres_table[genres_table['genre'] != '(no genres listed)'].reset_index(drop=True)
display(genres_table)

genre  movie_count
0        Action         7348
1     Adventure         4145
2     Animation         2929
3      Children         2935
4        Comedy        16870
5         Crime         5319
6   Documentary         5605
7         Drama        25606
8       Fantasy         2731
9     Film-Noir          353
10       Horror         5989
11         IMAX          195
12      Musical         1054
13      Mystery         2925
14      Romance         7719
15       Sci-Fi         3595
16     Thriller         8654
17          War         1874
18      Western         1399

In [138]:
for genre in genres_table['genre'].unique():
  movies_df[genre] = np.where(movies_df['genres'].str.contains(genre), 1, 0)
movies_df.head()

movieId                        title  \
0        1                    Toy Story   
1        2                      Jumanji   
2        3             Grumpier Old Men   
3        4            Waiting to Exhale   
4        5  Father of the Bride Part II   

                                        genres  year  Action  Adventure  \
0  Adventure|Animation|Children|Comedy|Fantasy  1995       0          1   
1                   Adventure|Children|Fantasy  1995       0          1   
2                               Comedy|Romance  1995       0          0   
3                         Comedy|Drama|Romance  1995       0          0   
4                                       Comedy  1995       0          0   

   Animation  Children  Comedy  Crime  Documentary  Drama  Fantasy  Film-Noir  \
0          1         1       1      0            0      0        1          0   
1          0         1       0      0            0      0        1          0   
2          0         0       1      0            0      0        0          0   
3          0         0       1      0            0      1        0          0   
4          0         0       1      0            0      0        0          0   

   Horror  IMAX  Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0       0     0        0        0        0       0         0    0        0  
1       0     0        0        0        0       0         0    0        0  
2       0     0        0        0        1       0         0    0        0  
3       0     0        0        0        1       0         0    0        0  
4       0     0        0        0        0       0         0    0        0

### Let's take a closer look at 'tags.csv' - this file contains every single tag action done by users:

In [223]:
tags = pd.read_csv(local_path+'/tags.csv')
tags.head()

userId  movieId               tag   timestamp
0       3      260           classic  1439472355
1       3      260            sci-fi  1439472256
2       4     1732       dark comedy  1573943598
3       4     1732    great dialogue  1573943604
4       4     7569  so bad it's good  1573943455

In [229]:
print('There are {} unique tags in the dataset'.format(len(tags['tag'].unique())))
tag_counts = tags[['tag', 'movieId']].groupby('tag').count().sort_values('tag').reset_index()
tag_counts.head(50)

There are 73051 unique tags in the dataset


tag  movieId
0                                 Alexander Skarsgård        1
1                                Difficult to find it        1
2                                      Filmes Antigos        2
3                                     Filmes Antigos         2
4                                       Kartik Aaryan        1
5                                         Kriti Sanon        1
6                                       Laurel Canyon        1
7                                       Luis Brandoni        1
8                                     Masami Nagasawa        1
9                                  O'Shea Jackson Jr.        1
10                               The Lost Boys Series        3
11                                      ballet school        1
12                                            breakup        1
13                                     mental illness        2
14                                   sherlock holmes         1
15                                           spin off        1
16                            !950's Superman TV show        1
17                                "A Christmas Carol"        1
18                                  "Duck and cover!"        2
19                                 "Ghost for adults"        1
20                               "I need a computer."        1
21                                       "Jack black"        1
22                    "Jessica Biel is hot like hell"        1
23                     "May"-weirdness -- not as good        1
24            "No high heels on the hardwood floors!"        2
25    "Nobody with a good car needs to be justified."        1
26                                "Nut up or shut up"        2
27                                           "Oh Yah"        1
28                               "Show me the money."        3
29                                       "The Hunter"        4
30                                            "Tonite        1
31                      "We're on a mission from god"        1
32                                            "artsy"        1
33                         "bad CIA"-- too simplistic        1
34                         "bad cia"-- too simplistic        1
35                             "based on true events"        2
36                                           "bitchy"        1
37                                          "comédie"        1
38                                  "damn dirty apes"       12
39                                 "eccentric genius"        1
40                                    "found footage"       80
41                                 "hero" is murderer        1
42                               "inspired" by a book        1
43     "it wasn't expecially large, but it felt hard"        1
44                       "jackie chan" "chris tucker"        1
45  "less than an hour of air" yeah more like 10 m...        1
46                                        "lost" film        1
47                           "magic negro" stereotype        1
48                               "meet me in montauk"        1
49       "middlebrow" attempt at "intelligent" ending        1

### As we can see, there's a lot of standard text cleansing processes we can do to make things a bit clearer, including stripping any spaces, getting rid of odd characters etc - let's do this now:

In [163]:
def string_cleanse(string):
  string_clean = str(string).strip().lower().replace('"','').replace('#','').replace('\'','').replace('(','').replace(')','').replace('*','').replace('-','')
  return string_clean
tags['tag'] = tags['tag'].apply(string_cleanse)
clean_tag_counts = tags[['tag', 'movieId']].groupby('tag').count().sort_values('tag').reset_index()
clean_tag_counts.head(50)

tag  movieId
0                                                           11
1                              !950s superman tv show        1
2                                           &suspense        1
3                                     +++++++++++++++        1
4                                     ,music business        1
5                                                   .        1
6   ...and this film has got nothing to with the a...        1
7                       ...livin in an amish paradise        1
8                            ...that its finally over        1
9                   ...to pay me to watch this again.        1
10                                                /b/        1
11                                          /suspense        1
12                                          0 gravity        1
13                                            0 stars        1
14                                                007      214
15                                007 movies are bad.        1
16                                         007 series       75
17                                            007like        1
18                                           01.02.05        1
19                                           01.03.06        1
20                                           01.04.06        1
21                                           01.10.05        1
22                                              01/10       77
23                                              01/11      167
24                                              01/12      113
25                                              01/14        1
26                                              01/15        5
27                                           02.04.06        1
28                                              02/10       54
29                                              02/11      159
30                                              02/14        1
31                                              02/15        5
32                                              03/09       27
33                                              03/10       90
34                                              03/11      170
35                                              03/15        7
36                                           04.03.06        1
37                                           04.04.06        1
38                                              04/09       10
39                                              04/10        6
40                                              04/11       95
41                                              04/14        1
42                                              04/15        5
43                                           05.02.06        1
44                                           05.03.06        1
45                                           05.04.06        1
46                                              05/09        6
47                                              05/10       83
48                                              05/11      168
49                                              05/13        1

Next let's start to get some counts of how often these tags appear in the dataset.

Each row of the dataset is a tag applied by a user to a film - let's filter to a single, popular film, to get a feel for what this looks like - I have chosen 2001: A Space Odyssey as it is one of my favourites and is also quite a popular film. This film has id '924'

In [233]:
chosen_film_id = 924
chosen_film_tags = tags[tags['movieId'] == chosen_film_id]
print('Film ID {} has {} total tags, of which {} are unique, from {} different users\n'.format(chosen_film_id, chosen_film_tags['tag'].count(), chosen_film_tags['tag'].nunique(), chosen_film_tags['userId'].nunique()))
chosen_film_tags['tag'].value_counts()

Film ID 924 has 1749 total tags, of which 191 are unique, from 323 different users



Stanley Kubrick                          105
artificial intelligence                  103
sci-fi                                    97
space                                     89
atmospheric                               84
philosophical                             77
visually appealing                        74
slow                                      72
surreal                                   70
masterpiece                               61
cinematography                            52
cult film                                 43
meditative                                40
confusing ending                          40
future                                    38
music                                     38
classic                                   37
futuristic                                35
aliens                                    31
space travel                              26
soundtrack                                26
boring                                    26
robots    

### Let's look at the top 50 most used tags:
### This however desn't account for the fact that certain tags might be repeatedly used for certain films - e.g. the tag 'Harry Potter' might be repeatedly used across all the Harry Potter franchise - but isn't all that useful for our recommender.

In [165]:
tags[['tag', 'movieId']].groupby('tag').count().sort_values('movieId', ascending=False).reset_index().head(50)

tag  movieId
0                scifi     9141
1          atmospheric     7053
2               action     6783
3               comedy     6368
4              surreal     5584
5                funny     5354
6      based on a book     5194
7         twist ending     4904
8   visually appealing     4691
9              romance     4482
10            dystopia     4329
11         dark comedy     4133
12              quirky     3999
13                 bdr     3966
14    thoughtprovoking     3848
15             fantasy     3818
16            stylized     3758
17             classic     3733
18          psychology     3646
19      woman director     3590
20      nudity topless     3590
21            violence     3386
22         time travel     3306
23              murder     3302
24           adventure     3288
25   social commentary     3251
26                dark     3248
27            thriller     3247
28           superhero     3246
29             revenge     3146
30               drama     3135
31               space     3078
32      cinematography     3076
33           animation     3043
34          disturbing     2945
35            suspense     2929
36              horror     2904
37              aliens     2877
38       coming of age     2777
39    great soundtrack     2660
40               music     2632
41               drugs     2602
42           criterion     2598
43       psychological     2579
44             musical     2570
45        black comedy     2501
46          true story     2486
47         predictable     2476
48          friendship     2475
49           cult film     2373

### There also seems to be some further text clean-up required - look at all the variations of 'sci-fi'!

In [166]:
temp = tags[['tag', 'movieId']].groupby('tag').count().sort_values('movieId', ascending=False).reset_index()
temp[temp['tag'].str.lower().str.contains('sci')].head(20)

tag  movieId
0                 scifi     9141
80      science fiction     1635
250             science      651
278         bad science      604
544           scientist      329
607       mad scientist      298
758       classic scifi      238
761              sci fi      237
1112            fascism      160
1279   thoughtful scifi      136
1381       good science      121
1543  intelligent scifi      106
1681      consciousness       96
1952         scifi cult       80
2159          joe pesci       70
2164         scientific       70
2249   computer science       66
2942       dumb science       48
2995      pseudoscience       47
3075     cerebral scifi       45

### Let's create a function which takes in a list of variant tags that we want to standardise to one tag...

In [158]:
def condenser(string, from_list, to_value):
  for i in from_list:
    string = str(string).replace(i, to_value)
  return string

###...and let's use it on our scifi variants!

In [168]:
scifi_variants = ['science fiction', 'sci fi']
tags['tag'] = tags['tag'].apply(condenser, from_list=scifi_variants, to_value='scifi')

### Here we can see that some of the most popular films can have hundreds of unique tags, while many have just 1

In [11]:
import numpy as np
unique_tags_by_movie = tags.groupby('movieId')['tag'].nunique().reset_index()
unique_tags_by_movie.columns = ['movieId', 'unique_tag_count']
unique_tags_by_movie.sort_values('unique_tag_count', ascending=False, inplace=True)
unique_tags_by_movie.reset_index(inplace=True, drop=True)
unique_tags_by_movie

movieId  unique_tag_count
0          260               698
1          356               521
2          296               498
3          318               384
4         1103               282
...        ...               ...
45246   158805                 1
45247   176869                 1
45248   176867                 1
45249   176865                 1
45250   180607                 1

[45251 rows x 2 columns]

### Let's create a frequency table of unique tag counts - we can see that 15% of our films have just 1 tag, while over half have 5 tags or less.

In [12]:
tagcount_freqtable = unique_tags_by_movie.groupby('unique_tag_count').count().reset_index()
tagcount_freqtable.columns = ['unique_tag_count', 'frequency']
tagcount_freqtable['percent'] = tagcount_freqtable['frequency']/(tagcount_freqtable['frequency'].sum())
tagcount_freqtable['cumulative_percent'] = tagcount_freqtable['percent'].cumsum()
display(tagcount_freqtable.head(20))

unique_tag_count  frequency   percent  cumulative_percent
0                  1       7047  0.155731            0.155731
1                  2       5379  0.118870            0.274602
2                  3       4919  0.108705            0.383306
3                  4       3976  0.087865            0.471172
4                  5       3228  0.071335            0.542507
5                  6       2561  0.056595            0.599103
6                  7       2002  0.044242            0.643345
7                  8       1694  0.037436            0.680781
8                  9       1280  0.028287            0.709067
9                 10       1165  0.025745            0.734812
10                11       1008  0.022276            0.757088
11                12        912  0.020154            0.777242
12                13        753  0.016641            0.793883
13                14        688  0.015204            0.809087
14                15        587  0.012972            0.822059
15                16        554  0.012243            0.834302
16                17        561  0.012398            0.846700
17                18        404  0.008928            0.855627
18                19        369  0.008155            0.863782
19                20        397  0.008773            0.872555

### Let's take a look at that film with 698 tags - id 260

### I think we can all guess what film this is!
### As we can see, there are a lot of meaningless tags ('bite me'), a lot of tags with character names ('yoda' - only really relevant to this particular film/franchise), a lot of tags all meaning pretty much the same thing ('science-fiction', 'sf,science fiction', 'Science Fiction', 'sci-fi') and a lot of free form opinions ('birth of great scifi ideas', 'best movie ever'

### We can also see that there is a lot of upper/lower case versions of similar things

### I think next we should aggregate the other way round, and look at how many tags are used just once/very infrequently

In [13]:
tags[tags['movieId'] == 260].tag.unique()

array(['classic', 'scifi', 'action', 'adventure', 'fantasy',
       'space adventure', 'classic scifi', 'good vs evil', 'aliens',
       'oldie but goodie', 'scifi cult', 'space', 'cult classic',
       'futuristic', 'space action', 'space opera', 'old movie', 'epic',
       'harrison ford', 'space epic', 'action comedy', 'romance',
       'entertaining', 'good story', 'imdb top 250', 'darth vader',
       'luke skywalker', 'the death star', 'religion', 'heros journey',
       'george lucas', 'star wars', 'epic adventure',
       'birth of great scifi ideas', 'old fx quality', 'story driven',
       'bite me',
       'episode what? its cut off, so i dont even know what movie it is',
       'action, scifi', 'quotable', 'exciting', 'fun', 'heroic journey',
       'inventive', 'amazing', 'masterpiece', 'must see', 'universe',
       'sf,scifi', 'incest', 'james earl jones', 'jedi', 'john williams',
       'robots', 'atmospheric', 'future', 'great soundtrack',
       'science fantasy', 'sp

### Here we have the top 20 tags in the dataset, ranked by how many unique films the tag is applied to.
### The top one here is 'bdr' which means 'Blu-ray Disc - Recordable' which is not actually anything to do with the film
### Same logic applies to 'betamax' and 'dvdvideo'
### However some of these tags are indeed useful

In [172]:
def calc_unique_movies_by_tag(tags_df, tag_col='tag', movie_col='movieId'):
  df_agg = tags.groupby(tag_col)[movie_col].nunique().reset_index()
  df_agg.columns = ['tag', 'unique_movie_count']
  df_agg.sort_values('unique_movie_count', ascending=False, inplace=True)
  df_agg.reset_index(inplace=True, drop=True)
  return df_agg

movie_count_by_tag = calc_unique_movies_by_tag(tags)
movie_count_by_tag.head(20)

tag  unique_movie_count
0                      bdr                3948
1           woman director                3493
2                   murder                2271
3         independent film                1838
4                   comedy                1624
5           nudity topless                1474
6          based on a book                1395
7                      clv                1357
8                    drama                1307
9                  romance                1296
10                   funny                1165
11                violence                1152
12  based on novel or book                1130
13                 revenge                1121
14                 musical                1117
15               criterion                1103
16                 betamax                1079
17                    love                1070
18                  family                1004
19                  action                 978

### Let's also construct a frequency table as before:
### An incredible 57.6% of tags (37,498) only apply to one film!

In [173]:
def calc_tag_freq_table(tags_df, tag_col='tag', movie_col='movieId'):
  df_agg = calc_unique_movies_by_tag(tags_df, tag_col, movie_col)
  df_freq = df_agg.groupby('unique_movie_count').count().reset_index()
  df_freq.columns = ['unique_movie_count', 'frequency']
  df_freq['percent'] = df_freq['frequency']/(df_freq['frequency'].sum())
  df_freq['cumulative_percent'] = df_freq['percent'].cumsum()
  return df_freq

moviecount_freqtable = calc_tag_freq_table(tags)
display(moviecount_freqtable.head(20))

unique_movie_count  frequency   percent  cumulative_percent
0                    1      37498  0.576059            0.576059
1                    2       8345  0.128199            0.704258
2                    3       4157  0.063861            0.768120
3                    4       2604  0.040004            0.808124
4                    5       1795  0.027576            0.835699
5                    6       1402  0.021538            0.857237
6                    7       1027  0.015777            0.873014
7                    8        821  0.012613            0.885627
8                    9        633  0.009724            0.895351
9                   10        539  0.008280            0.903632
10                  11        487  0.007481            0.911113
11                  12        416  0.006391            0.917504
12                  13        349  0.005361            0.922865
13                  14        291  0.004470            0.927336
14                  15        282  0.004332            0.931668
15                  16        268  0.004117            0.935785
16                  17        203  0.003119            0.938904
17                  18        181  0.002781            0.941684
18                  19        179  0.002750            0.944434
19                  20        181  0.002781            0.947215

### Creating a list of tags that apply to just one film and scanning through, we can see that the quality of these are not particularly useful

In [16]:
movie_count_by_tag[movie_count_by_tag['unique_movie_count']==1]['tag'].to_list()

['yihad',
 'the virgin suicides',
 'tray',
 'wicked stepmother',
 'travelling players',
 'what the hell did i just watch?',
 'this movie is perfect',
 'travis tope',
 'polemic',
 'the voiceovers here arent as corny as people say',
 'plant food',
 'theological debate',
 'plc hacks',
 'theeffects',
 'psycholgical thriller',
 'theranos',
 'wrongful sentence',
 'transcending abilities',
 'transcendental',
 'pol',
 'the villain from karate kid at the bar',
 'travis oates',
 'yellowstone national park',
 'this movie scares the stuffing out of me!',
 'what are you stuffed with?',
 'playwright:william gibson',
 'plot light',
 'plot point:photo booth',
 'track star',
 'tracking',
 'theology student',
 'this is what they say the notebook was',
 'travis cluff',
 'theone',
 'yegor letov',
 'tracking camera',
 'psychobiddy',
 'transcendentialist',
 'track down',
 'tracking shot',
 'psychobabble',
 'yiddish writer',
 'the views',
 'provoking',
 'yigal amir',
 'this movie is pretty...pretty gay',
 'p

Let's take only the tags that apply to more than 100 films
As previously discussed, there are some formats in here which we can immediately erase, such as 'bdr' (bluray disc), 'clv' (laserdisc), 'betamax', 'dvdvideo', 'bdvideo' 'turneysdvds' etc, as well as some other undesirable tags

In [175]:
movie_count_by_tag = calc_unique_movies_by_tag(tags)
movie_count_by_tag.head(250)

tag  unique_movie_count
0                            bdr                3948
1                 woman director                3493
2                         murder                2271
3               independent film                1838
4                         comedy                1624
5                 nudity topless                1474
6                based on a book                1395
7                            clv                1357
8                          drama                1307
9                        romance                1296
10                         funny                1165
11                      violence                1152
12        based on novel or book                1130
13                       revenge                1121
14                       musical                1117
15                     criterion                1103
16                       betamax                1079
17                          love                1070
18                        family                1004
19                        action                 978
20                      dvdvideo                 953
21                        boring                 949
22                     biography                 936
23                    friendship                 930
24                           sex                 897
25                      suspense                 860
26                         music                 842
27                        sequel                 824
28                         drugs                 805
29                        police                 799
30                         short                 778
31                   documentary                 751
32                         crime                 722
33                      thriller                 713
34                       bdvideo                 706
35                         scifi                 703
36                        horror                 702
37         less than 300 ratings                 694
38                   tumeys dvds                 673
39                    might like                 664
40                   atmospheric                 659
41                  world war ii                 648
42                   predictable                 626
43                           war                 618
44                 coming of age                 617
45             adapted from:book                 616
46                     animation                 614
47                         sport                 611
48     parent child relationship                 609
49                 serial killer                 602
50         based on a true story                 587
51                      politics                 587
52                           gay                 579
53                          rape                 577
54                       surreal                 576
55                 new york city                 568
56                   high school                 559
57                        dvdram                 556
58                       fantasy                 556
59                         death                 552
60                    kidnapping                 552
61                      teenager                 549
62                        acting                 549
63                       history                 534
64                cinematography                 530
65                       classic                 529
66                      religion                 528
67                        remake                 519
68                    soundtrack                 514
69                        prison                 514
70                     adventure                 511
71                       suicide                 508
72                          slow                 508
73                  martial arts                 504
74                     cult film                 498
7

### Let's create a helper function to explore different tags:

In [180]:
def tag_explorer(tag, tags_df, movies_df):
  temp = tags_df[tags_df['tag'] == tag][['movieId', 'tag', 'userId']]
  temp2 = temp.groupby(['movieId', 'tag'], as_index=False).count()
  temp2.sort_values('userId', ascending=False, inplace=True)
  temp2.columns = ['movieId', 'tag', 'user_count']
  temp2.reset_index(drop=True, inplace=True)
  temp2 = temp2.merge(movies_df[['movieId', 'title']], how='left', on='movieId')
  temp3 = temp2.reindex(['title', 'tag', 'user_count'], axis=1)
  return print(temp3.head(20))

In [222]:
temp = tags[tags['tag'] == 'surreal']#[['movieId', 'tag', 'userId']]
temp[temp['movieId']==7147]

userId  movieId      tag   timestamp tag_condenser_test
509         93     7147  surreal  1496543859            surreal
510         93     7147  surreal  1496543872         surrealism
4192      1541     7147  surreal  1433455810            surreal
4193      1541     7147  surreal  1433455818         surrealism
7625      2730     7147  surreal  1450767760            surreal
7626      2730     7147  surreal  1450767764         surrealism
19350     4117     7147  surreal  1442297742            surreal
218777    7335     7147  surreal  1502931705            surreal
218778    7335     7147  surreal  1502931712         surrealism
259550   15204     7147  surreal  1306924418            surreal
259551   15204     7147  surreal  1555780024         surrealism
262635   15321     7147  surreal  1291402658            surreal
262636   15321     7147  surreal  1291402660         surrealism
273431   17348     7147  surreal  1527014019            surreal
273432   17348     7147  surreal  1527014040         surrealism
275991   18839     7147  surreal  1535136700            surreal
275992   18839     7147  surreal  1535136709         surrealism
279514   19346     7147  surreal  1534083196            surreal
279515   19346     7147  surreal  1534083205         surrealism
297238   21096     7147  surreal  1240889655            surreal
320543   24693     7147  surreal  1227772942            surreal
320544   24693     7147  surreal  1227772972         surrealism
323038   25512     7147  surreal  1262486296            surreal
323039   25512     7147  surreal  1262486295         surrealism
326926   26054     7147  surreal  1252640312         surrealism
356618   31371     7147  surreal  1296475021            surreal
384135   36282     7147  surreal  1161141022         surrealism
386135   36692     7147  surreal  1491065038            surreal
386940   37329     7147  surreal  1370734729         surrealism
388318   38203     7147  surreal  1466543082            surreal
399480   41285     7147  surreal  1314908975            surreal
409356   43069     7147  surreal  1249855609         surrealism
444899   49315     7147  surreal  1435857020            surreal
455509   51750     7147  surreal  1377056288            surreal
458531   51803     7147  surreal  1498411998            surreal
458532   51803     7147  surreal  1498412041         surrealism
462439   53613     7147  surreal  1311879022            surreal
462440   53613     7147  surreal  1311879025         surrealism
469516   54526     7147  surreal  1284300189            surreal
469517   54526     7147  surreal  1284300195         surrealism
476553   56152     7147  surreal  1440289340            surreal
476554   56152     7147  surreal  1394470047         surrealism
487850   58084     7147  surreal  1332532426            surreal
487851   58084     7147  surreal  1332532415         surrealism
489484   58632     7147  surreal  1139273418         surrealism
505810   61718     7147  surreal  1455985008            surreal
505811   61718     7147  surreal  1455985014         surrealism
538539   65068     7147  surreal  1360560091            surreal
552833   67731     7147  surreal  1280886306            surreal
578692   75351     7147  surreal  1263225179            surreal
580557   75702     7147  surreal  1163802857         surrealism
583783   77076     7147  surreal  1292036509            surreal
583784   77076     7147  surreal  1292036511         surrealism
585837   77965     7147  surreal  1567659462            surreal
605274   81185     7147  surreal  1525690360            surreal
605275   81185     7147  surreal  1525690366         surrealism
633515   86780     7147  surreal  1564557281            surreal
633516   86780     7147  surreal  1564557287         surrealism
640378   88819     7147  surreal  1156990481         surrealism
641315   89307     7147  surreal  1530397874            surreal
641795   89320     7147  surreal  1147412719         surrealism
654352   92430     7147  surreal  145178

In [221]:
temp.groupby('movieId').count().loc[7147,:]

tag       100
userId    100
Name: 7147, dtype: int64

### The tag 'r' has no discernable information behind it:

In [181]:
tag_explorer('r', tags, movies_df)

                                                title tag  user_count
0   Borat: Cultural Learnings of America for Make ...   r           8
1                               Letters from Iwo Jima   r           5
2                  Before the Devil Knows You're Dead   r           5
3                                           In Bruges   r           5
4                               Science of Sleep, The   r           5
5                                       Departed, The   r           5
6                      Blind Swordsman: Zatoichi, The   r           5
7                                      Twelve Monkeys   r           5
8                                            Hot Fuzz   r           5
9                    Girl with the Dragon Tattoo, The   r           5
10                              Thank You for Smoking   r           4
11                                            Paprika   r           4
12                                     Gone Baby Gone   r           4
13                  

### ridiculous is a matter of opinion and probably not that useful:

In [182]:
tag_explorer('ridiculous', tags, movies_df)

                                         title         tag  user_count
0                                  Pacific Rim  ridiculous          23
1                    Grand Budapest Hotel, The  ridiculous          20
2                                    Kung Fury  ridiculous          18
3                                        Signs  ridiculous          16
4                                     Face/Off  ridiculous          13
5                 Kingsman: The Secret Service  ridiculous          12
6           Monty Python's The Meaning of Life  ridiculous          10
7                                 Shoot 'Em Up  ridiculous          10
8                               The Babysitter  ridiculous          10
9                                    True Lies  ridiculous           9
10                             Jennifer's Body  ridiculous           9
11               You Don't Mess with the Zohan  ridiculous           8
12         Transformers: Revenge of the Fallen  ridiculous           8
13  Tw

### The 'too long' tag, whle I may agree with some of these, is not really something I think anyone would ask for in a movie recommendation!

In [190]:
tag_explorer('too long', tags, movies_df)

                                     title       tag  user_count
0                        The Hateful Eight  too long          32
1                                   Avatar  too long          29
2       Hobbit: An Unexpected Journey, The  too long          28
3                           Shutter Island  too long          27
4                         Django Unchained  too long          21
5     Hobbit: The Desolation of Smaug, The  too long          21
6                Blue Is the Warmest Color  too long          19
7                                 Watchmen  too long          19
8                               Mr. Nobody  too long          17
9                              Cloud Atlas  too long          17
10                          Enter the Void  too long          15
11  Lord of the Rings: The Two Towers, The  too long          14
12                                  Zodiac  too long          13
13                               Boot, Das  too long          13
14                       

### Just to finish on a positive, I'd definitely agree with the surreal tag - it shows that these tags are indeed going to be useful!

In [189]:
tag_explorer('surreal', tags, movies_df)

                                    title      tag  user_count
0   Eternal Sunshine of the Spotless Mind  surreal         225
1                               Inception  surreal         196
2                            Donnie Darko  surreal         175
3                              Fight Club  surreal         159
4                        Mulholland Drive  surreal         124
5                         Pan's Labyrinth  surreal         113
6                                  Amelie  surreal         112
7                             Matrix, The  surreal         102
8                    Being John Malkovich  surreal          95
9                              Mr. Nobody  surreal          93
10                              Dark City  surreal          72
11                                 Brazil  surreal          71
12                  2001: A Space Odyssey  surreal          70
13                           Annihilation  surreal          68
14                               Coraline  surreal     

### Let's gather all the usless tags we've discovered above, and remove them from the tags dataframe:

In [207]:
format_tags = ['bdr', 'clv', 'betamax', 'dvdvideo', 'bdvideo', 'tumeys dvds', 'less than 300 ratings', 'dvd', 'dvdr', 'dvdram', 'vhs']
unknown_tags = ['prospect preferred', 'r']
undesirable_tags = ['rape', 'suicide', 'nudity full frontal', 'sex', 'torture', 'incest', 'nudity', 'prostitution', 'nudity topless', 'nudity topless  brief', 'nudity full frontal  notable', 'child abuse', 'nudity rear', 'nudity topless  notable']
genre_tags = ['scifi', 'action', 'comedy', 'romance', 'fantasy', 'adventure', 'thriller', 'drama', 'animation', 'horror', 'musical', 'crime', 'mystery', 'war', 'documentary', 'animated', 'british comedy']
time_tags = ['1930s', '1950s', '1960s', '1970s', '1980s', '1990s']
useless_opinion_tags = ['seen more than once', 'movie to see', 'reviewed', 'to see', 'ridiculous', 'boring', 'overrated', 'might like']
deletion_lists = [format_tags, unknown_tags, undesirable_tags, genre_tags, time_tags, useless_opinion_tags]
tag_deletions = [tag for tag_list in deletion_lists for tag in tag_list]
tags = tags[~tags['tag'].isin(tag_deletions)].reset_index(drop=True)

### Let's also clean up some of the similar ones in the top 500:

In [208]:
literature_variants = ['based on a book', 'based on novel or book', 'adapted from:book']
tags['tag'] = tags['tag'].apply(condenser, from_list=literature_variants, to_value='book adaptation')

philosophy_variants = ['philosophical', 'philosophy']
tags['tag'] = tags['tag'].apply(condenser, from_list=philosophy_variants, to_value='philosophy')

nyc_variants = ['new york city', 'new york']
tags['tag'] = tags['tag'].apply(condenser, from_list=nyc_variants, to_value='new york')

surreal_variants = ['surreal', 'surrealism']
tags['tag'] = tags['tag'].apply(condenser, from_list=surreal_variants, to_value='surreal')

alien_variants = ['alien', 'aliens']
tags['tag'] = tags['tag'].apply(condenser, from_list=alien_variants, to_value='alien')

cult_variants = ['cult classic', 'cult', 'cult film']
tags['tag'] = tags['tag'].apply(condenser, from_list=cult_variants, to_value='cult')

dance_variants = ['dance', 'dancing']
tags['tag'] = tags['tag'].apply(condenser, from_list=dance_variants, to_value='dance')

antihero_variants = ['antihero', 'dark hero']
tags['tag'] = tags['tag'].apply(condenser, from_list=antihero_variants, to_value='antihero')

In [209]:
top500tags = tags['tag'].value_counts().head(500)
top500tags = pd.DataFrame(top500tags).reset_index()
top500tags.columns = ['tag', 'tag_count']
top500tags.sort_values('tag')

tag  tag_count
315                             19th century        474
439                                       3d        359
282                                     70mm        520
279                                   absurd        532
143                                   acting        933
438                             adam sandler        360
478                       adapted from:comic        331
379                                addiction        408
205                                 adultery        667
258                                   africa        564
255                                al pacino        572
457                               alcoholism        347
326                         alfred hitchcock        461
14                                     alien       3519
233                           alien invasion        620
65                         alternate reality       1619
327                                  amnesia        459
494                           angelina jolie        321
455                                  animals        348
68                                     anime       1585
488                            anne hathaway        326
377                          anthony hopkins        409
193                                 antihero        706
180                               apocalypse        739
294                    arnold schwarzenegger        505
469                                      art        340
55                   artificial intelligence       1801
423                                 artistic        373
132                                 assassin        974
1                                atmospheric       7053
344                                australia        443
147                               bad acting        903
462                                 bad plot        343
238                              bad science        604
339                         based on a comic        447
347                          based on a play        441
39                     based on a true story       2351
368                       based on a tv show        416
449                           based on comic        352
338                                   batman        448
76                                 beautiful       1441
333                        beautiful scenery        455
272                       beautifully filmed        542
251                        bechdel test:fail        578
243                        bechdel test:pass        594
447                              ben affleck        353
177                              bill murray        743
60                                 biography       1713
44                               bittersweet       2131
81                           black and white       1395
33                              black comedy       2501
342                                    bleak        444
316                                    blood        471
465                                bollywood        342
0                            book adaptation       7575
361                                   boxing        424
108                                brad pitt       1097
89                                   british       1236
127                             bruce willis        999
311                                   brutal        476
350                                 business        440
226                                    campy        630
331                              cannibalism        456
236                                car chase        607
302                                     cars        489
235                                 cerebral        613
352                                      cgi        434
201                                   cheesy        675
199                              chick flick        697
407                                childhood        385
176                                 children        743
412      